From 59d650bab0a9ddc31797b8277a8d7752ff0bdd84 Mon Sep 17 00:00:00 2001
From: Chih-Wei Huang <cwhuang@linux.org.tw>
Date: Thu, 18 Jul 2019 15:32:13 +0800
Subject: [PATCH 24/24] Fix forceDefaultOrientation

Ignore screen size when determines mForceDefaultOrientation.

Fix the bug in the calculation of left and top of DisplayContent
when default orientation is forced. (by Ladehunter)
---
 .../java/com/android/server/policy/PhoneWindowManager.java   | 2 +-
 services/core/java/com/android/server/wm/DisplayContent.java | 5 +++--
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/policy/PhoneWindowManager.java b/services/core/java/com/android/server/policy/PhoneWindowManager.java
index 56ceda74c6b5..8d635602ff74 100644
--- a/services/core/java/com/android/server/policy/PhoneWindowManager.java
+++ b/services/core/java/com/android/server/policy/PhoneWindowManager.java
@@ -2419,7 +2419,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
         // so if the orientation is forced, we need to respect that no matter what.
         final boolean isTv = mContext.getPackageManager().hasSystemFeature(
                 PackageManager.FEATURE_LEANBACK);
-        mForceDefaultOrientation = ((longSizeDp >= 960 && shortSizeDp >= 720) || isCar || isTv) &&
+        mForceDefaultOrientation = /* ((longSizeDp >= 960 && shortSizeDp >= 720) || isCar || isTv) && */
                 res.getBoolean(com.android.internal.R.bool.config_forceDefaultOrientation) &&
                 // For debug purposes the next line turns this feature off with:
                 // $ adb shell setprop config.override_forced_orient true
diff --git a/services/core/java/com/android/server/wm/DisplayContent.java b/services/core/java/com/android/server/wm/DisplayContent.java
index 22ac65df4337..3829c356acae 100644
--- a/services/core/java/com/android/server/wm/DisplayContent.java
+++ b/services/core/java/com/android/server/wm/DisplayContent.java
@@ -2952,12 +2952,13 @@ class DisplayContent extends WindowContainer<DisplayContent.DisplayChildWindowCo
         // Uses same calculation as in LogicalDisplay#configureDisplayInTransactionLocked.
         final int orientation = mDisplayInfo.rotation;
         boolean rotated = (orientation == ROTATION_90 || orientation == ROTATION_270);
+        boolean forcedOrientation = mService.mPolicy.isDefaultOrientationForced();
         final int physWidth = rotated ? mBaseDisplayHeight : mBaseDisplayWidth;
         final int physHeight = rotated ? mBaseDisplayWidth : mBaseDisplayHeight;
         int width = mDisplayInfo.logicalWidth;
-        int left = (physWidth - width) / 2;
+        int left = forcedOrientation ? 0 : (physWidth - width) / 2;
         int height = mDisplayInfo.logicalHeight;
-        int top = (physHeight - height) / 2;
+        int top = forcedOrientation ? 0 : (physHeight - height) / 2;
         out.set(left, top, left + width, top + height);
     }
 
-- 
2.17.1

